### ISSUE TITLE
switch build system?

### ISSUE THREAD

**Issue by stsp** [dosemu2/fdpp#146]
Currently (in master) fdpp is only
buildable on focal and groovy.
Bionic misses llvm-objcopy so
that binutils is needed, and Xenial
misses lld, so binutils is needed.
There are a few ways to solve that:
1. Revert back to objcopy(binutils)
and forget about Xenial and about
llvm-only build.
2. Revert back to objcopy+ld.bfd
(binutils) and forget about non-binutils
builds forever.
3. Switch to xmake and do
non-binutils builds when possible.
4. Create fdpp-legacy
5. Do nothing

Case 3 can only work if you can,
as @andrewbird suggests, say
"lld | binutils" in Build-Depends.

I would personally think I'd go for
1 unless @jschwartzenberg moves
things to 3. :)

**Comment by andrewbird** [comment]
I can swap Travis to use focal for both fdpp and dosemu2, which is probably the right thing to do now anyway.

Option 4 seems to align quite nicely with dosemu-legacy.
Option 3 is my least favourite, as I'd prefer that parallel builds worked sensibly with make, but of course I don't have to maintain the build system.

Locally I'm still using Bionic simply because the Lubuntu on Focal is so awful after its switch to the Qt tookit, that it was unusable! I'm probably going to need to switch desktops to Xubuntu which breaks my heart as Lubuntu on Bionic is just right for me.

**Comment by stsp** [comment]
I don't prefer 4 here because,
unlike in dosemu2, where old
builds reduce both security and
functionality, here nothing is
changed at all. So such separation
here makes much less sense
than there.

make was never a problem here
(unlike in dosemu2), so I am not
a big fan of 3 too, but some configure
step may be helpful, as it is turning
out now:
- bfd/lld/gold
- nasm/yasm
- llvm-objcopy/objcopy
- clang/gcc (not any soon)
- sed/gsed (see parse_decls.sh where I probe for gsed by hands)
- various flavours of `install` (worked to a common denominator already)

Build options:
- debug/non-debug/extra-debug
- use/unuse sanitizers

I'd say, the list is not as small as
it was when I thought "oh, 10-lines
makefile will suffice".

**Comment by stsp** [comment]
I reverted to objcopy to get at
least Bionic back.
But I still think its time to get
configure, as too many build-time
uncertainty have accumulated.

**Comment by minexew** [comment]
@stsp why do you write
everything
as a Haiku?

It's quite difficult
to parse.

**Comment by stsp** [comment]
I think you see the automatic
line-wrapping laying on top of
the manual ones. I am trying
to keep the lines short to avoid
automatic wraps from ever
happening, but sometimes they
happen nevertheless. Resulting in
something
like this. :)

**Comment by jschwartzenberg** [comment]
I'm not sure that happened, I think @minexew is just generally curious about your automatic-wrapping-fobia ;)

**Comment by stsp** [comment]
I think its time to push this forward
again, as it seems we'll need to switch
from lld to GNU ld, which doesn't support
the cross-compilation for free. I.e. lld
supports all targets at once, whereas
GNU ld needs to be installed in a special
cross-build form.
I looked into xmake docs, and it says
is has a package management of its own,
as well as supporting the native package
managers. apt is listed, but rpm/dnf - not.
@waruqi are you planning the dnf/rpm
support?
Also I wonder how exactly I can benefit
from that package management. For
the LP/COPR work-flows I think it would
be very cool if xmake could just generate
the rpkg and debuild specs automatically,
with all the needed deps.
Basically I think this is more about an
integration with native builders, rather
with package managers. I.e. it would be
cool if xmake could, besides installing
deps and building the package on its
own, also generate the needed spec,
on which debuild/rpkg can be used.
Can it do something like this?

**Comment by waruqi** [comment]
>  but rpm/dnf - not.
@waruqi are you planning the dnf/rpm
support?

It may be considered in the future, and you can open a feature request first. 
We now support other package managers in addition to apt, such as conan/vcpkg/pacman/zypper/portage.

> Basically I think this is more about an
integration with native builders, rather
with package managers. I.e. it would be
cool if xmake could, besides installing
deps and building the package on its
own, also generate the needed spec,
on which debuild/rpkg can be used.
Can it do something like this?

like cmake/cpack? We are also planning to implement a similar feature to generate deb/nsis/rpm packages, but it may take some time, maybe a few more releases. Next release we will start with an initial implementation of support for nsis packages. Then we'll move on to support other packages like deb, rpm, etc.

BTW, xmake has been in the debian and archlinux repository.  https://packages.debian.org/sid/xmake

**Comment by stsp** [comment]
Generating packages is probably not
what I need. I am looking for generating
only the specs for lp/copr/obs work-flows.
Is something like that planned?

**Comment by waruqi** [comment]
> Generating packages is probably not what I need. I am looking for generating only the specs for lp/copr/obs work-flows. Is something like that planned?

I don't understand the format of these files. Maybe you can say more detail, or you can generate them yourself by way of a custom task.

**Comment by stsp** [comment]
rpkg: https://github.com/dosemu2/fdpp/blob/master/fdpp.spec.rpkg
deb: https://github.com/dosemu2/fdpp/blob/master/debian/control
obs: https://build.opensuse.org/package/view_file/home:stsp2/dosemu2/dosemu2.spec?expand=1

I of course can generate them by
hands, but:
1. the build system has all the needed info for them
2. whenever the dep is changed, I need to update 4 places!

**Comment by stsp** [comment]
> We are also planning to implement a similar feature to generate deb/nsis/rpm packages

Note that you will very likely need
to generate these specs anyway, if
you want to build the appropriate
package. Because the build tools
need it. Of course unless you somehow
can build them by xmake "directly",
which I don't think is possible, or?

**Comment by stsp** [comment]
https://github.com/mesonbuild/meson/issues/49
meson seems to be doing that
since 2015.

**Comment by waruqi** [comment]
> > We are also planning to implement a similar feature to generate deb/nsis/rpm packages
> 
> Note that you will very likely need to generate these specs anyway, if you want to build the appropriate package. Because the build tools need it. Of course unless you somehow can build packages by xmake "directly", bypassing the host tools. Which I don't think is possible, or?

Yes, I meant to generate these specs for packaging. xmake doesn't generate them directly. It generates all kinds of specs and then calls debuild, nsismake and other packaging tools to generate it. In addition, it does better automation, such as automatically installing required tools(debuild, nsismake ...) and then calling them.

Of course, we can also tell xmake to only generate specs and not to auto-package them by using the policy configuration.

> [mesonbuild/meson#49](https://github.com/mesonbuild/meson/issues/49) meson seems to be doing that since 2015.

We also have a related issue:

https://github.com/xmake-io/xmake/issues/1433

The custom api feature in the recent dev release is actually in preparation for this packaging feature https://github.com/xmake-io/xmake/issues/4276

And we're already planning to implement the initial nsis spec generation in 2.8.6, as well as some initial design of the packaging and configuration APIs. If all goes well, and I have enough time, I will start trying to implement deb and rpm spec generation and packaging in 2.8.7. Of course, I haven't decided which one to implement first.

**Comment by stsp** [comment]
Thanks, this all sounds promising.
The only problem is that something like
5 years ago I was getting the same
answers, and there is afaics still no
xmake in debian or fedora... :)

**Comment by stsp** [comment]
Also you likely know that there is a
small difference between rpm and rpkg
specs. Copr uses rpkg, obs uses rpm,
but rpm spec can be generated from
rpkg spec, so it might be better to generate
rpkg, not rpm spec.

**Comment by waruqi** [comment]
> Thanks, this all sounds promising. The only problem is that something like 5 years ago I was getting the same answers, and there is afaics still no xmake in debian or fedora... :)

We've been working on historical issues and we just recently resolved a 2 year old issue. https://github.com/xmake-io/xmake/issues/1452

It will take more time because I'm the only one maintaining it.

Also, xmake has recently been in debian. https://packages.debian.org/sid/xmake

By the time ubuntu 24.04 is released, we can get it.

**Comment by waruqi** [comment]
> Also you likely know that there is a small difference between rpm and rpkg specs. Copr uses rpkg, obs uses rpm, but rpm spec can be generated from rpkg spec, so it might be better to generate rpkg, not rpm spec.

We'll deal with those differences when we start working on this.

**Comment by stsp** [comment]
24.04 ok, any ETA for fedora?

**Comment by stsp** [comment]
Ah, you mentioned fedora already, sorry.

**Comment by stsp** [comment]
You don't have Mantic builds on LP.
Why is so?

**Comment by waruqi** [comment]
> You don't have Mantic builds on LP. Why is so?

xmake was just uploaded to the debian repositories last week, and ubuntu 23.10 has been frozen.

**Comment by stsp** [comment]
I mean this:
https://launchpad.net/~xmake-io/+archive/ubuntu/xmake
Can you add Mantic build?

**Comment by waruqi** [comment]
> I mean this: https://launchpad.net/~xmake-io/+archive/ubuntu/xmake Can you add Mantic build?

I haven't added it to the ppa package yet, it will be included in the next xmake release.

**Comment by stsp** [comment]
I don't understand.
I enable the new builds with 1 mouse
click. What's the use waiting for the
next release, to just enable the build
on LP?
I simply can't apt-get it right now.
Why to wait?

**Comment by waruqi** [comment]
> I don't understand. I enable the new builds with 1 mouse click. What's the use waiting for the next release, to just enable the build on LP? I simply can't apt-get it right now. Why to wait?

I don't know how to enable it on the web fastly, at the moment I can only upload it via ci when a new version is released. https://github.com/xmake-io/xmake/blob/042c51144c1a918d75493a94094e29aeafd924ff/.github/ workflows/linux.yml#L50

**Comment by stsp** [comment]
Hmm, not sure.
Today I went to LP UI and enabled
the "Noble" build of fdpp by just checking
in the checkbox. I always did so before, too.
I can't give you any hints, as the only thing
I know to do on LP, is to mouse-click through
their convoluted UIs.
I am currently uploading my nasm fork there,
finding the click-through ways, and am enabling
the needed builds w/o problems.

**Comment by waruqi** [comment]
> Hmm, not sure. Today I went to LP UI and enabled the "Noble" build of fdpp by just checking in the checkbox. I always did so before, too. I can't give you any hints, as the only thing I know to do on LP, is to mouse-click through their convoluted UIs. I am currently uploading my nasm fork there, finding the click-through ways, and am enabling the needed builds w/o problems.

I didn't find the corresponding checkbox, but I've uploaded it by modifying the script to use a temporary ci to upload it. You can look at it again.

**Comment by stsp** [comment]
![изображение](https://github.com/dosemu2/fdpp/assets/3295627/66d0d886-104a-4033-91c0-74e48e204734)

**Comment by stsp** [comment]
Let me show you an UI, its here. :)

**Comment by waruqi** [comment]
> Let me show you an UI, its here. :)

Thanks, but I still can't find it, I don't know where the daliy-build page is.

<img width="1432" alt="image" src="https://github.com/dosemu2/fdpp/assets/151335/a7ecfb79-bf2b-404b-b855-d890b2073540">
<img width="1432" alt="image" src="https://github.com/dosemu2/fdpp/assets/151335/cfaf772e-2766-4c8b-bdcc-2ca20be1b790">

**Comment by stsp** [comment]
Its in "Code" section, not "Overview".

**Comment by waruqi** [comment]
> Its in "Code" section, not "Overview".

nothing

<img width="1433" alt="image" src="https://github.com/dosemu2/fdpp/assets/151335/ad9a779e-b5de-46c7-bc11-6c8c5e7d9961">
<img width="1142" alt="image" src="https://github.com/dosemu2/fdpp/assets/151335/2698f35d-5878-4a68-a1e0-9b68843e75de">

**Comment by stsp** [comment]
I think you need to create a "series"
(trunk series), create a source import,
and then create a package recipe.
Every time I am doing this for another
project, I spending hours for finding
the correct click-through path. Its
always a quest. But at the end it
usually works. :)

**Comment by waruqi** [comment]
Oh, sounds a bit complicated, I don't have much time to work on this. mantic I've uploaded it via script and it's pending, wait a bit more and you should be able to download it.

<img width="1289" alt="image" src="https://github.com/dosemu2/fdpp/assets/151335/cb1819d8-50a6-4675-8a5c-2a7499e60cbb">

**Comment by stsp** [comment]
Yes, downloaded, thanks.
Btw, are there any writings that
compare xmake to others?
Yes, I've seen your own, but for
me its too vague. Not very concrete,
not enough examples, and too terse.
Maybe there is something else to read?
My problem is that I am not quite
familiar with pretty much any build
system, other than make and autotools.
So I can't even formalize my needs and
say "this system has what I need, and
this doesn't". Because I don't know what
I need. :) Maybe reading a comparisons
will help me to set up my mind straight.

**Comment by waruqi** [comment]
> Yes, downloaded, thanks. Btw, are there any writings that compare xmake to others? Yes, I've seen your own, but for me its too vague. Not very concrete, not enough examples, and too terse. Maybe there is something else to read? My problem is that I am not quite familiar with pretty much any build system, other than make and autotools. So I can't even formalize my needs and say "this system has what I need, and this doesn't". Because I don't know what I need. :) Maybe reading a comparisons will help me to set up my mind straight.

Yes, most of the articles are mine. You can read some or docs.

- https://github.com/xmake-io/xmake/wiki/C-and-Cplusplus-build-system,-I-use-xmake
- https://[xmake](https://xmake.io/#/getting_started).io/#/getting_started

But there are also some articles and videos written by others:

- https://www.youtube.com/watch?v=ar3muou4BQM

**Comment by stsp** [comment]
When I run xmake inside dosemu2
folder, it keeps silence for about 5sec,
and then says:
```
note: configure found, try building it or you can run `xmake f --trybuild=` to set buildsystem (pass -y or --confirm=y/n/d to skip confirm)?
please input: y (y/n)
```

5 seconds for nothing??
Come on!

**Comment by waruqi** [comment]
> When I run xmake inside dosemu2 folder, it keeps silence for about 5sec, and then says:
> 
> ```
> note: configure found, try building it or you can run `xmake f --trybuild=` to set buildsystem (pass -y or --confirm=y/n/d to skip confirm)?
> please input: y (y/n)
> ```
> 
> 5 seconds for nothing?? Come on!

If your project does not have any xmake.lua files, xmake will do some of the following things

1. It tries to detect if the current project matches any existing packages in the xmake-repo repository, and if it does, it tries to build it using its package configuration. If it matches, it will use its package configuration to try to build it. As long as it matches exactly, the success rate of such a build is almost 100%, and xmake will automatically patch any known build errors. Detection is a bit time-consuming because it has to load all the package configuration scripts, but on my machine it only takes 2.8s!

```lua
ruki-2:dosemu2 ruki$ time xmake
note: configure.ac found, try building it or you can run `xmake f --trybuild=` to set buildsystem
 (pass -y or --confirm=y/n/d to skip confirm)?
please input: y (y/n)
^C

real    0m2.884s
user    0m0.655s
sys     0m0.346s
```

2. If it doesn't match the corresponding package script, then it will continue probing known build systems (autotools, cmake ...) ) to attempt a build. This has a certain failure rate, but it usually simplifies a lot for the user, and it also helps the user to simplify cmake/autoconf compilation for cross-compilation, android/iphone, etc.

**Comment by stsp** [comment]
But I think silently spending 3-5 sec
is unacceptable, no matter what actually
happens. Given that its a "first start",
it can produce quite a bad first impression.
I am sure you can remove that lag,
keeping all the features in.

**Comment by stsp** [comment]
And what if I don't have internet
connection? Will it spend even more
time trying to contact its repo?

**Comment by waruqi** [comment]
> And what if I don't have internet connection? Will it spend even more time trying to contact its repo?

will not

**Comment by waruqi** [comment]
> But I think silently spending 3-5 sec is unacceptable, no matter what actually happens. Given that its a "first start", it can produce quite a bad first impression. I am sure you can remove that lag, keeping all the features in.

I have improved it. you can try 

```lua
xmake update dev
source ~/.xmake/profile
```

**Comment by stsp** [comment]
Thanks, I'll wait for a PPA update.
Can you enable Noble PPA build in
a mean time?

**Comment by waruqi** [comment]
> Thanks, I'll wait for a PPA update. Can you enable Noble PPA build in a mean time?

I'll enable noble when it's released. Also, you should be able to download xmake directly from noble, even without using ppa.

**Comment by stsp** [comment]
> Also, you should be able to download xmake directly from noble, even without using ppa.

Is it already possible?

**Comment by waruqi** [comment]
> > Also, you should be able to download xmake directly from noble, even without using ppa.
> 
> Is it already possible?

I think so, it's already in the debian repositories.

**Comment by stsp** [comment]
https://launchpad.net/ubuntu/noble/amd64/xmake/2.8.3+ds-2
Seems so, congrats. :)

**Comment by stsp** [comment]
On some systems, like android,
there are lots of missing deps.
Also as now I am using the patched
build of nasm, most of every system
will get a missing dep.
So I wonder if xmake can deal with
that task: I can provide it with PPA
and copr repo of the needed deps,
so it can try to install them natively.
But on the systems where this doesn't
apply, it should build the deps from
sources. I guess for that I'll need to
set up some repo with xmake recipes
for such deps.
What automation level does xmake
provide for such scenario? Can it
automatically choice from the list
of repos, which ones are natively
applicable to install the certain deps,
and which are needed as a fall-back
to build the dep in-place?

**Comment by waruqi** [comment]
If the dependencies have been committed to the xmake-repo repository and it supports android, e.g., zlib, libpng, etc., then you can use it. Then we simply need to configure it.

```lua
add_requires("zlib")
add_requires("libpng 1.6.x")

target("test")
    add_files("src/*.c")
    add_packages("libpng", "zlib")
```

xmake will download their source tar package, then build and install it automatically.

we need only run

```console
$ xmake f -p android
$ xmake
```

like this:

```console
ruki-2:test ruki$ xmake f -p android --ndk=~/files/android-ndk-r20b/
checking for architecture ... armeabi-v7a
checking for Android SDK directory ... ~/Library/Android/sdk
checking for Build Tools Version of Android SDK ... 33.0.1
checking for NDK directory ... /Users/ruki/files/android-ndk-r20b
checking for SDK version of NDK ... 21
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> zlib v1.3 [from:libpng]
  -> libpng v1.6.40 
please input: y (y/n/m)

  => install zlib v1.3 .. ok         
  => download https://mirrors.ustc.edu.cn/debian/pool/main/libp/libpng1.6/libpng1.6_1.6.40.orig.tar.gz .. ok
  => install libpng v1.6.40 .. ok      

ruki-2:test ruki$ xmake
[ 25%]: cache compiling.release src/main.cpp
[ 50%]: linking.release test
[100%]: build ok, spent 0.172s
```

**Comment by waruqi** [comment]
Also, if they are already installed on the system, xmake will preferentially find and use them from the system via pkg-config, or you can configure `add_requires("apt::zlib-dev")` to integrate it directly from an apt installation if an apt source exists.

**Comment by stsp** [comment]
> If the dependencies have been committed to the xmake-repo repository

No, can I instead put them into the
source tree itself? Like I do with m4
macros for autoconf.

> Also, if they are already installed on the system

What if they are not, but are available
via dnf/apt? Although perhaps this is a
bad case, since xmake would need root
to install those...

**Comment by waruqi** [comment]
> > If the dependencies have been committed to the xmake-repo repository
> 
> No, can I instead put them into the source tree itself? Like I do with m4 macros for autoconf.

put 3rd library source code to self project? xmake supports it too. see

https://xmake.io/#/package/local_3rd_source_library

> 
> > Also, if they are already installed on the system
> 
> What if they are not, but are available via dnf/apt? Although perhaps this is a bad case, since xmake would need root to install those...

Unless you explicitly configure `add_requires("apt::xxx")`, xmake will go ahead and call apt to install it, and it will request the user root. I wouldn't normally recommend this either.

With `add_requires("zlib")`, xmake will just look for the system libraries, and if it doesn't find them, it will download them from the xmake-repo repository, but it won't call apt.

**Comment by stsp** [comment]
> put 3rd library source code to self project?

No, just the xmake build recipes, for
the case when they are missing on a
target system.

**Comment by waruqi** [comment]
? I don't understand your need. I haven't used m4 much.

**Comment by stsp** [comment]
> If the dependencies have been committed to the xmake-repo repository

But what I want, is not to commit
anything into xmake-repo. I want
to put xmake recipes into my project,
how to build various deps (for which
the xmake recipes do not exist). And
I'd like xmake, if not found in already
installed packages, to d/l sources for
these deps (using the URLs that I need
to specify somewhere), and use the
supplied recipes to build them.

**Comment by waruqi** [comment]
> > If the dependencies have been committed to the xmake-repo repository
> 
> But what I want, is not to commit anything into xmake-repo. I want to put xmake recipes into my project, how to build various deps (for which the xmake recipes do not exist). And I'd like xmake, if not found in already installed packages, to d/l sources for these deps (using the URLs that I need to specify somewhere), and use the supplied recipes to build them.

Then you just need to put the package configuration definitions from the xmake-repo into your project's xmake.lua .

You can put your package to your project or your local repository

see https://xmake.io/#/package/remote_package?id=using-self-built-private-package-repository

```
projectdir
    - myrepo
      - packages
        - z/zlib/xmake.lua
```

```lua
add_repositories("my-repo myrepo")

add_requires("zlib")
```

or does not repositories

```lua
package("zlib")
    ...
    on_install(...)
package_end()

add_requires("zlib")

target("test")
    add_packages("zlib")
```

**Comment by stsp** [comment]
Yeah, private per-project repo should
be enough!
So to summarize:
```
projectdir
    - myrepo
      - packages
        - z/zlib/xmake.lua
```
By doing so, I'll be able to only
put zlib/xmake.lua into my project,
this xmake.lua will contain the URL
for source download (I don't want
to put zlib sources neither in my
project nor in any remote repo),
and xmake will use that optionally,
given zlib is not installed.
Correct? :)

**Comment by waruqi** [comment]
> Yeah, private per-project repo should be enough! So to summarize:
> 
> ```
> projectdir
>     - myrepo
>       - packages
>         - z/zlib/xmake.lua
> ```
> 
> By doing so, I'll be able to only put zlib/xmake.lua into my project, this xmake.lua will contain the URL for source download (I don't want to put zlib sources neither in my project nor in any remote repo), and xmake will use that optionally, given zlib is not installed. Correct? :)

right, you can only set remote url of zlib source.

**Comment by stsp** [comment]
OK, sounds like this can indeed
boost the cross-platform portability.
As currently the only thing we can
do (with autoconf) is to write
"this package is missing, please install it"

**Comment by stsp** [comment]
OK, I think I need some kind of the
recursive layout to start the conversion.
The projects are too large, and the
builds are too scripted, to start
converting from top-level.
But I can easily start from converting
some small sub-targets. Initially I can
invoke xmake from make for those
sub-targets. But I wonder how those
sub-targets would share the global
configure parameters set by xmake?
Is there somewhere a guide on writing
the xmake recipes incrementally for
the large project?

**Comment by stsp** [comment]
> sub-targets would share the global
> configure parameters set by xmake?

Besides configure parameters, the
question is also about build dir. If
every sub-target is built into its own
build dir, things won't work. Somehow
all sub-targets need to respect the
build dir of the whole project.

**Comment by waruqi** [comment]
> OK, I think I need some kind of the recursive layout to start the conversion. The projects are too large, and the builds are too scripted, to start converting from top-level. But I can easily start from converting some small sub-targets. Initially I can invoke xmake from make for those sub-targets. But I wonder how those sub-targets would share the global configure parameters set by xmake? Is there somewhere a guide on writing the xmake recipes incrementally for the large project?

share global xmake configure parameters to make? 

If you want to share configuration in xmake.lua, you can use `includes()` api.

see https://xmake.io/#/guide/syntax_description?id=multi-level-configuration

**Comment by stsp** [comment]
> If you want to share configuration in xmake.lua, you can use includes() api.

But how is to get that working with
the otherwise make-based system?
So for example I want to convert just
one sub-target. According to your
link, I need the top-level xmake.lua
which will include the component's
xmake.lua. But since the entire project
is still built by make, I'll need to build
the converted component from its
sub-dir (rather than from a top dir),
and make sure its build directory
matches the make's build directory,
rather than to follow the xmake's
build dir layout. Usually this means:
build this component in _current_
directory, locating its sources via
the command-line parameter.
Is this possible?

Something like this:
`xmake --build-dir=. --src-dir=$(SRCDIR)`
in a makefile.
Is this possible?

**Comment by waruqi** [comment]
> > If you want to share configuration in xmake.lua, you can use includes() api.
> 
> But how is to get that working with the otherwise make-based system? So for example I want to convert just one sub-target. According to your link, I need the top-level xmake.lua which will include the component's xmake.lua. But since the entire project is still built by make, I'll need to build the converted component from its sub-dir (rather than from a top dir), and make sure its build directory matches the make's build directory, rather than to follow the xmake's build dir layout. Usually this means: build this component in _current_ directory, locating its sources via the command-line parameter. Is this possible?
> 
> Something like this: `xmake --build-dir=. --src-dir=$(SRCDIR)` in a makefile. Is this possible?

see https://github.com/xmake-io/xmake/issues/3342

```console
$ xmake f -P $(SRCDIR)-o $(BUILDIR)
$ xmake
```

**Comment by stsp** [comment]
Of course the main problem seems
to be the sparse docs... But thanks for
sitting here, fulfilling that gap yourself. :)

**Comment by stsp** [comment]
> $ xmake f -P $(SRCDIR) -o $(BUILDIR)

Doesn't seem to work as expected.
```
$ xmake f -o .
checking for platform ... linux
checking for architecture ... x86_64
$ xmake 
[ 20%]: compiling.yacc thunk_gen.y
[ 40%]: compiling.lex thunk_gen.l
[ 60%]: linking.release thunk_gen
[100%]: build ok, spent 0.263s
```

I wanted to get `./thunk_gen`, but instead
I am getting `linux/x86_64/release/thunk_gen`.
Also I didn't want to set the output
dir with a separate command, that
does some "configuration". I just
wanted to set the output dir with
the cmdline switch, to easily integrate
the xmake's component into a
makefile-based project.

**Comment by waruqi** [comment]
> > $ xmake f -P (SRCDIR)−o(BUILDIR)
> 
> Doesn't seem to work as expected.
> 
> ```
> $ xmake f -o .
> checking for platform ... linux
> checking for architecture ... x86_64
> $ xmake 
> [ 20%]: compiling.yacc thunk_gen.y
> [ 40%]: compiling.lex thunk_gen.l
> [ 60%]: linking.release thunk_gen
> [100%]: build ok, spent 0.263s
> ```
> 
> I wanted to get `./thunk_gen`, but instead I am getting `linux/x86_64/release/thunk_gen`. Also I didn't want to set the output dir with a separate command, that does some "configuration". I just wanted to set the output dir with the cmdline switch, to easily integrate the xmake's component into a makefile-based project.

```
set_targetdir("$(buildir)")
```

**Comment by stsp** [comment]
With that I am getting `build/thunk_gen`.
And if I do `set_targetdir(".")` then running
xmake from source dir gives me `./thunk_gen`.
But running it from builddir this way:
`xmake -P /home/stas/src/fdpp/fdpp/parsers`
gives me `/home/stas/src/fdpp/fdpp/parsers/thunk_gen`
rather than `./thunk_gen` in builddir.
Maybe I should set target dir via option,
rather than writing it into the xmake.lua?

**Comment by waruqi** [comment]
> With that I am getting `build/thunk_gen`. And if I do `set_targetdir(".")` then running xmake from source dir gives me `./thunk_gen`. But running it from builddir this way: `xmake -P /home/stas/src/fdpp/fdpp/parsers` gives me `/home/stas/src/fdpp/fdpp/parsers/thunk_gen` rather than `./thunk_gen` in builddir. Maybe I should set target dir via option, rather than writing it into the xmake.lua?

you can define custom option. https://xmake.io/#/manual/configuration_option?id=defining-options

```lua
option("outputdir", {description = "the output directory"})

target("test")
    set_targetdir("$(outputdir)")
```

```console
xmake f --outputdir=xxxx
```

**Comment by waruqi** [comment]
> Is the typo in "buildir" intentional?

history reason, I don't like double `d`.

> Can I set it without an extra step, whatever this "f" is?

you need config it first. `xmake f/config --xxx=`, then run `xmake build` command.

**Comment by stsp** [comment]
Can you think of some config-build
combo mode for better integration with
existing build systems, where the
config step was already done behind
xmake's back?
In such mode the user provides as
much as he wants via cmdline, the
rest gets defaulted, and the build
process starts.
What do you think?

**Comment by stsp** [comment]
No, quite the opposite.
I am converting make-based build
system to xmake target-by-target.
This have nothing to do with trybuild.
Instead, I have all configuration params
already in a variables. So I just need
to pass them to xmake, bypassing
its own config step. Like this:
`xmake -P /home/stas/src/fdpp/fdpp/parsers --buildir=.`
In this example I want to pass the
buildir to it without doing 2 invocations.

**Comment by waruqi** [comment]
> No, quite the opposite. I am converting make-based build system to xmake target-by-target. This have nothing to do with trybuild. Instead, I have all configuration params already in a variables. So I just need to pass them to xmake, bypassing its own config step. Like this: `xmake -P /home/stas/src/fdpp/fdpp/parsers --buildir=.` In this example I want to pass the buildir to it without doing 2 invocations.

you can define custom option. `xmake f --foo=xxx --bar=xxx`

> > With that I am getting `build/thunk_gen`. And if I do `set_targetdir(".")` then running xmake from source dir gives me `./thunk_gen`. But running it from builddir this way: `xmake -P /home/stas/src/fdpp/fdpp/parsers` gives me `/home/stas/src/fdpp/fdpp/parsers/thunk_gen` rather than `./thunk_gen` in builddir. Maybe I should set target dir via option, rather than writing it into the xmake.lua?
> 
> you can define custom option. https://xmake.io/#/manual/configuration_option?id=defining-options
> 
> ```lua
> option("outputdir", {description = "the output directory"})
> 
> target("test")
>     set_targetdir("$(outputdir)")
> ```
> 
> ```
> xmake f --outputdir=xxxx
> ```

**Comment by stsp** [comment]
Yes, but this still requires 2 invocations.
I've got that to work, yes.
But what I propose is to think up
something to avoid 2 invocations,
because the configuration step was
already done by another build system.

**Comment by stsp** [comment]
So when xmake runs on behalf of
some other build system, maybe
you can think up the way to bypass
its own config step and instead get
all the needed params from the parent
build system.

**Comment by waruqi** [comment]
> Yes, but this still requires 2 invocations. I've got that to work, yes. But what I propose is to think up something to avoid 2 invocations, because the configuration step was already done by another build system.

You can't avoid the `xmake config` call, it's needed for all xmake tasks such as `xmake build`, `xmake run`, `xmake install`. It is not possible to pass arguments repeatedly for each of these tasks individually. Even configure/cmake/meson requires a separate configuration task.

**Comment by stsp** [comment]
Why not to have the combo mode
config+build, that can do 2 things
with 1 call?

**Comment by waruqi** [comment]
As I just said, there are too many combinations such as config+build, config+build+run, config+build+install . It is impossible to define a large number of repetitive arguments for each combination of commands, it is very unmaintainable.

**Comment by waruqi** [comment]
On unix/posix systems, it is very lightweight to perform two process calls, so why does it have to be one? With `xmake f -o output; xmake build` and `xmake build -o ouput`, there's not much difference between them, and only a few characters less. But the maintainability advantage of a standalone config task is obvious.

**Comment by stsp** [comment]
Firstly, I have no idea what is "f". :)
Secondly, my build system currently
does:
`cd builddir && $(MAKE) $@`
for every sub-target.
It then translates to:
`cd parsers && xmake f -o . -P $(abspath $(srcdir))/parsers && srcdir=$(abspath $(srcdir))/parsers xmake`

which is, as you can see, quite clumsy
and full of duplication. I probably can
get rid of `srcdir` here. Currently I do
```
    local srcdir = os.getenv("srcdir")
    if srcdir == nil then
        srcdir = "."
    end
    add_includedirs(srcdir)
```

**Comment by stsp** [comment]
How can I create the target for
the generated source files? The
closest I could find in your docs
is asn1, but it still builds the binary
by pre-defined rules.
What I need is completely custom
rules to produce various source
files. Are there the examples of this?
`make` has the special notion for
such "intermediate" targets.

**Comment by waruqi** [comment]
> How can I create the target for the generated source files? The closest I could find in your docs is asn1, but it still builds the binary by pre-defined rules. What I need is completely custom rules to produce various source files. Are there the examples of this? `make` has the special notion for such "intermediate" targets.

autogen? 

https://github.com/xmake-io/xmake/blob/master/tests/projects/other/autogen_codedep/xmake.lua
https://github.com/xmake-io/xmake/blob/master/tests/projects/other/autogen_code/xmake.lua

**Comment by stsp** [comment]
Not quite.
Yes, these examples demonstrate the
custom rule, but still they do
`set_kind("binary")` and produce the
binary at the end.
I also do need binary at the end,
obviously, but don't forget that I am
integrating into an existing build
system.
So can I do `set_kind("generated_cpp")`
or alike, while the binary is produced
by the parent build system?

**Comment by stsp** [comment]
I mean, the concept of intermediate
targets. Even make has that.

**Comment by waruqi** [comment]
> Not quite. Yes, these examples demonstrate the custom rule, but still they do `set_kind("binary")` and produce the binary at the end. I also do need binary at the end, obviously, but don't forget that I am integrating into an existing build system. So can I do `set_kind("generated_cpp")` or alike, while the binary is produced by the parent build system?

other build system generate cpp and call xmake to build it?

you can set  always_added to add these generated cpp files. `add_files("$(buildir)/autogen.cpp", {always_added = true})`
https://github.com/xmake-io/xmake/blob/master/tests/projects/other/autogen_code/xmake.lua

or in your custom rule, you can call compiler to compile these generated files.

**Comment by stsp** [comment]
> other build system generate cpp and call xmake to build it?

Opposite: xmake generates them,
parent build system build on them.
For that I envision something like
`set_kind("generated_cpp")`.

**Comment by waruqi** [comment]
> > other build system generate cpp and call xmake to build it?
> 
> Opposite: xmake generates them, parent build system build on them. For that I envision something like .`set_kind("generated_cpp")`

You can generate it in many ways

### phony target + custom script

```lua
target("test")
    set_kind("phony")
    on_config(function (target)
       -- generate cpp
    end)
```

or in before_build, ...

2. binary + on_build script

```lua
target("test")
    set_kind("binary")
    on_build(function (target)
       -- generate cpp
    end)
```

it will override the whole build stage. you can ignore binary kind.

3. target +  custom rule

4. custom task

https://xmake.io/#/manual/plugin_task

you can define a custom task to generate cpp, then call it in makefile

```lua
target("generate_cpp")
  -- TODO
```

```console
xmake generate_cpp
```

5. custom lua script

https://xmake.io/#/plugin/builtin_plugins?id=run-the-custom-lua-script

/home/generate_cpp.lua

```lua
function main()
    -- generate cpp
end
```

```console
xmake lua /home/generate_cpp.lua
```

**Comment by stsp** [comment]
While there are 1024+ ways to
generate anything, either with
or without xmake, what intermediate
targets should do, is at least to
handle the updates properly.
That is, when some target A depends
on the intermediate target B, then
any change of pre-requisites of B
should re-build B and A. No changes
of pre-requisites of B should avoid
re-building of B unnecessarily.
So if I do `xmake generate_cpp`,
as you suggest above, I guess it
would just call my generator func
which will re-build everything
unnecessarily.

So what of the proposed ways
actually treat the generated results
as a target, with proper deps tracking?

**Comment by waruqi** [comment]
You can add dependency files to detect if it needs to be regenerated. https://github.com/xmake-io/xmake/blob/3bdfa2178d729709c40669ec9f2e361636f01eb3/tests/projects/other/autogen_codedep/xmake.lua#L15

> Yes, these examples demonstrate the
custom rule, but still they do
and produce the
binary at the end.

you can remove batchcmds:compile to only generate cpp. https://github.com/xmake-io/xmake/blob/3bdfa2178d729709c40669ec9f2e361636f01eb3/tests/projects/other/autogen_codedep/xmake.lua#L12

> as you suggest above, I guess it
would just call my generator func
which will re-build everything
unnecessarily.xmake generate_cpp

you can also use `depend.on_changed()` in custom task/scripts to check dep/changed.

https://github.com/xmake-io/xmake/blob/3bdfa2178d729709c40669ec9f2e361636f01eb3/xmake/rules/verilator/verilator.lua#L218

**Comment by stsp** [comment]
> You can add dependency files to detect if it needs to be regenerated.

Thanks.

> you can remove batchcmds:compile to only generate cpp.

But what to do with set_kind()  in that case?

> you can also use depend.on_changed() in custom task/scripts to check dep/changed.

Can I get this more automated?
Instead of me copy-pasting the
low-level xmake code that I don't
understand (even if its quite obvious
how to modify it, as you suggest),
would it be possible for me to only
define the conversion rule, define
the intermediate target that can
apply that rule, and have xmake
to handle the rest, like proper
"xmake clean", on-demand updates,
and so on? Can I avoid duplicating
that code?

**Comment by waruqi** [comment]
> But what to do with set_kind() in that case?

If you set binary kind, it performs the default link behaviour. If you don't want linking, you can override on_build/on_link or set it to object kind to skip link stage. 

> Can I get this more automated?
Instead of me copy-pasting the
low-level xmake code that I don't
understand (even if its quite obvious
how to modify it, as you suggest),
would it be possible for me to only
define the conversion rule, define
the intermediate target that can
apply that rule, and have xmake
to handle the rest, like proper
"xmake clean", on-demand updates,
and so on? Can I avoid duplicating
that code?

Usually you just need to add the dependency file

https://github.com/xmake-io/xmake/blob/3bdfa2178d729709c40669ec9f2e361636f01eb3/tests/projects/other/autogen_codedep/xmake.lua#L15

```lua
rule("autogen")
    set_extensions(".in")
    on_buildcmd_file(function (target, batchcmds, sourcefile, opt)
        local sourcefile_cx = path.join(target:autogendir(), "rules", "autogen", path.basename(sourcefile) .. ".cpp")
        batchcmds:show_progress(opt.progress, "${color.build.object}autogen %s", sourcefile)
        batchcmds:mkdir(path.directory(sourcefile_cx))
        batchcmds:vrunv("echo", {sourcefile, sourcefile_cx})
        batchcmds:add_depfiles(sourcefile)
    end)

target("autogen")
    set_kind("object")
    add_files("src/*.in")
    add_rules("autogen")
```

**Comment by stsp** [comment]
> set_kind("object")

But why "object", if its a text file?

> Usually you just need to add the dependency file

Is this enough to handle `xmake clean`
correctly? From your example its not
clear to me how xmake would find out
which files were generated, to clean them
up properly.
Also I am not entirely sure how to add
such generate files as a deps for another
target.

**Comment by waruqi** [comment]
> But why "object", if its a text file?

only object, phony, binary, headeronly kinds. I don't want to add more kinds.

> Is this enough to handle xmake clean
correctly? From your example its not
clear to me how xmake would find out
which files were generated, to clean them
up properly.
Also I am not entirely sure how to add
such generated files as a deps for another
target.

you can add your custom cleanfiles . `add_cleanfiles()` in description scope. `target:add("cleanfiles", "xxx")` in script scope.

or you can define custom `on_clean` scripts.

**Comment by stsp** [comment]
So how can I add such generated
files as a deps for another target?
In the parallel build process, unless
xmake tracks them properly, they
may not yet be generated when
their dep is started to be built.

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/blob/8eefa03ee2cadf0d89126e18173c1e179bebc3c5/tests/projects/other/autogen_codedep/xmake.lua#L34
https://xmake.io/#/guide/build_policies?id=buildacross_targets_in_parallel

**Comment by stsp** [comment]
Oh...
So is it correct to assume that
the intermediate target support -
I mean, the proper one, without
disabling parallel build, adding
custom clean hooks and so on -
is not something to ever expect?

**Comment by waruqi** [comment]
> without
disabling parallel build

This policy does not completely disable parallel builds; source compilation inside the target is still parallel.

**Comment by stsp** [comment]
https://github.com/xmake-io/xmake/discussions/2831
I think this tells something similar
to what I get, complaining about
set_kind(). But I think he misses the
fact that secondary targets are not
supported, so allowing user-defined
kinds is likely still not enough to solve
the problem.

**Comment by stsp** [comment]
```
$ xmake f

note: the following packages were not found in any repository (check if they are spelled correctly):
  -> nasm-segelf

$ dpkg-query -s nasm-segelf
Package: nasm-segelf
Status: install ok installed
```

Why doesn't it see the installed package?

**Comment by waruqi** [comment]
> ```
> $ xmake f
> 
> note: the following packages were not found in any repository (check if they are spelled correctly):
>   -> nasm-segelf
> 
> $ dpkg-query -s nasm-segelf
> Package: nasm-segelf
> Status: install ok installed
> ```
> 
> Why doesn't it see the installed package?

Since this package was not added to xmake-repo and `apt::nasm-segelf` was not specified, xmake can only treat it as unknown and run `nasm-segelf --version` to try to detect it, which is unreliable.

To detect it more reliably, you need to add it to the xmake-repo, or define `package("nasm-segelf")` in your own repository, and then provide the on_fetch script, for example:

```lua
package("nasm-segelf")
    set_kind("toolchain")
    on_fetch(function (package, opt)
        if opt.system then
            return package:find_tool("nasm")
        end
    end)
    --[[ optional, if you want to install from source url
    on_install(function (package)
    end)]]
package_end()

add_requires("nasm-segelf")

target("test")
   add_files("src/*.S")
   set_toolchains("nasm@nasm-segelf")
```

BTW, Currently, xmake's nasm toolchain does not support automatic binding of custom nasm packages. It will always use the system's nasm, and the xmake nasm toolchain needs to be improved if you want the above configuration to work completely and reliably.

Currently, only the llvm/mingw/zig toolchains support automatic package binding and switching at compile time.

**Comment by stsp** [comment]
> Since this package was not added to xmake-repo and apt::nasm-segelf was not specified

Even doing `add_requires("apt::nasm-segelf", {system = true})`
still gives this:
```
note: the following packages were not found on your system, try again after installing them:
  -> apt::nasm-segelf latest
```

> xmake can only treat it as unknown and run nasm-segelf --version to try to detect it,
> which is unreliable.

But for me it could work.
If I could tell xmake an alias name to do
`alias-name --version", then things will
work, because it will then do `nasm --version`
and get "nasm-segelf" in a version string.
Although of course if should just detect
it via native package manager.

> To detect it more reliably, you need to add it to the xmake-repo, or define
> package("nasm-segelf")

No, you said, it can detect it locally on a
system. If this wasn't said, I wouldn't
be trying xmake now.

Also: how can I disable this silly
`updating repositories .. ok` at the first
start, that takes quite a lot of time?
Well, if I can't stop xmake from always
going to the internet over my back, I'll
just delete it and will advice everyone to
do the same. :) This is completely
intolerable, so I hope you can just tell
me how to disable that behavior forever. :)

> It will always use the system's nasm

I don't care, because, until it supports
intermediate targets, I won't port the
build system to it. I will just use it as a
supplement, to get and build custom
nasm. But the parent build system will
use it. So this is not a problem.

**Comment by waruqi** [comment]
> No, you said, it can detect it locally on a
system. If this wasn't said, I wouldn't
be trying xmake now.

in your xmake.lua 

```lua
package("nasm-segelf")
    set_kind("toolchain")
    on_fetch(function (package, opt)
        if opt.system then
            return package:find_tool("nasm")
        end
    end)
package_end()

add_requires("nasm-segelf", {system = true})
```

It only detect it from local system, and your package configuration is in your local project.

**Comment by waruqi** [comment]
>  This is completely
intolerable, so I hope you can just tell
me how to disable that behavior forever. :)updating repositories .. ok

```console
xmake g --network=private
```

**Comment by stsp** [comment]
```
$ xmake g --network=private
configure
{
    proxy_pac = pac.lua
    theme = default
    network = private
}
```
Why does it say that?
Anyway, works as expected.
Also the initial delay in unsupported
sources is now gone.
Can I include such behavior into the
project, so that even when building it
on another machine, it won't go to the
internet until I explicitly ask to download
the nasm sources?

> in your xmake.lua

No, this way it also finds regular nasm.
It doesn't look for "nasm-segelf" in a
"--version" output, neither does it look
via the apt/dpkg.
So... what is this all about?

**Comment by waruqi** [comment]
> Can I include such behavior into the
project, so that even when building it
on another machine, it won't go to the
internet until I explicitly ask to download
the nasm sources?

you need use dev version, `set_policy("network.mode", "private")` in your xmake.lua

```console
xmake update dev
source ~/.xmake/profile
```

**Comment by stsp** [comment]
> you need use dev version

I am using PPA version, is this a dev version?

**Comment by waruqi** [comment]
> No, this way it also finds regular nasm.
It doesn't look for "nasm-segelf" in a
"--version" output, neither does it look
via the apt/dpkg.

you can run custom command to check.

```lua
package("nasm-segelf")
    set_kind("toolchain")
    on_fetch(function (package, opt)
        if opt.system then
	    local ok = try {function () os.run("dpkg-query -s nasm-segelf"); return true end}
	    if ok then
                return {name = "nasm-segelf", program = "<binary path>", version = "1.0"}
	    end
        end
    end)
```

BTw , It tried `--version` after installing nasm-segelf. it's not found.

```
$ nasm-segelf --version
bash: nasm-segelf: command not found
```

**Comment by waruqi** [comment]
> > you need use dev version
> 
> I am using PPA version, is this a dev version?

no, dev branch version, because I just added it today.

**Comment by stsp** [comment]
> you can run custom command to check.

So is this what you "meant" saying
it supports querying the native package
manager?

> BTw , It tried --version after installing nasm-segelf. it's not found.

Yes, but if it checked the actual
version string returned by
`nasm --version`, it could find
"nasm-segelf" there. But for that
it needs to get a tool name and a
string to grep, which probably can't
be provided now.

> no, dev branch version, because I just added it today.

Will wait for PPA, not a problem.

**Comment by stsp** [comment]
> local ok = try {function () os.run("dpkg-query -s nasm-segelf"); return true end}

Of course you realize this is not the
way to query "native package manager",
as it may be rpm or whatever else.

**Comment by waruqi** [comment]
> > local ok = try {function () os.run("dpkg-query -s nasm-segelf"); return true end}
> 
> Of course you realize this is not the way to query "native package manager", as it may be rpm or whatever else.

it's only a example, you can run any commands to detect it.

```console
os.run("dpkg-query -s nasm-segelf")
os.run("nasm-segelf --version")
os.run("nasm-segelf --help")
-- find it from rpm 
```

**Comment by stsp** [comment]
Is this what you "meant" saying
it supports querying the native package
manager?

**Comment by waruqi** [comment]
> Is this what you "meant" saying it supports querying the native package manager?

If it is a library, then `add_requires("fmt", {system = true})` will automatically call pkg-config to look up the system libraries, and if its package is configured with apt's extsources, it will also look it up from apt/pacman.

https://github.com/xmake-io/xmake-repo/blob/0166c715ff86093f3fff717e5752f5e2cb2a8831/packages/f/fmt/xmake.lua#L30

But if it's only a binary program, like your nasm-pkgself

We just need to try running `nasm-segself --version` to detect it from the system. Or go further and detect it from apt/rpm in the on_fetch custom script.

In the current version, there is no support for natively looking up binaries from apt/rpm.

**Comment by stsp** [comment]
> In the current version, there is no support for natively looking up binaries from apt/rpm.

Any plans?

**Comment by waruqi** [comment]
> > In the current version, there is no support for natively looking up binaries from apt/rpm.
> 
> Any plans?

you can open a feature request first, I'll consider improvements in the future, but there's no guarantee that I will be supported in the next version. I still have a lot of issues to deal with.

You can try runing the custom command approach first I mentioned earlier to detect it. Or keep waiting.

**Comment by stsp** [comment]
Well, I can work around the problem
by adding a nasm-segelf symlink into
a package.
So lets do just that for now...

**Comment by stsp** [comment]
Didn't help!
```
$ nasm-segelf --version
nasm-segelf version 2.17rc0
$ xmake f
note: the following packages were not found in any repository (check if they are spelled correctly):
  -> nasm-segelf
```

... with `add_requires("nasm-segelf")` in xmake.lua.

**Comment by stsp** [comment]
OK, since xmake failed to automate basically
every task I tried to put on it, I resorted
to the classical "plug existing holes one-by-one"
approach.
I slimmed down nasm-segelf packaging
to not conflict with nasm, added a homebrew
recipe for it, and will probably just add
autoconf to patch up the remaining holes.
... like those that  homebrew is mostly for
Mac, and therefore of course also doesn't
talk to the native package managers...
But at least it works, and almost looks
like the only workable solution among
the currently existing abundance of package
managers. I only had to submit just a couple
of pull requests to its repo to get it running,
and they are already approved, so let's say
it took just 1 day.

**Comment by waruqi** [comment]
The xmake package has been in Fedora official repository. https://src.fedoraproject.org/rpms/xmake

And I have supported to pack NSIS package. https://github.com/xmake-io/xmake/issues/1433

I will continue to pack deb/rpm packages in the future.

**Comment by stsp** [comment]
This is good to know.
But w/o the intermediate target support
it has little use at porting the existing
complex build system on. Plus its intention
to always go to the internet "to read some
news and watch youtube" is unacceptable
for me personally, and produces the
multi-seconds delays on start up.
And this is besides the fact that I wasn't
able to get it to detect nasm-segelf, even
after providing such symlink, so that
`nasm-segelf --version` could work.

In contrast, I was quite impressed with
the homebrew manager (but I can't say
I tried the xmake's package management
already, so can't really compare).
So I think it would be best if I return to xmake
a few years later to see if its all still the same,
and then decide.

**Comment by stsp** [comment]
This is to add that, according to you,
custom nasm replacement is not yet
supported, native package manager
query is not yet supported, native
package build is not yet supported...

But I don't think I'll easily find some
other build system that allows me to
gradually port the project, target-by-target.
I think most of the existing build systems
are not designed to co-exist in another
build system and cooperate with it.
When our project at work faced the same
problem, they did a research and found
that only bazel was adequate for their
porting needs. Then they spent a few
man-years porting to bazel...
But I do have neither bazel, nor a few
man-years to spend on this.

**Comment by stsp** [comment]
It appears that meson supports
intermediate targets, generators
and any weird dependency propagation.
Or maybe it doesn't: it just gives you
an OOP DSL, on top of which you can
code up most of anything...
But the cost is huge.
Took me 2 full weeks to just get to the
barely working POC.
Given that dosemu2 is much larger than
fdpp, it will likely never be ported.
The amount of work is ridiculously high.

**Comment by stsp** [comment]
meson support got the lucky hash
commit though: a999999
Perhaps this means, it worth the
troubles after all. :)

**Comment by andrewbird** [comment]
> Would you like to play with it a bit and
let me know what you think?

Sure. I never used meson before myself, so how should I test the build?

**Comment by stsp** [comment]
What command does it execute?
Can you add `set -x` to a script?

**Comment by andrewbird** [comment]
```diff
ajb@calypso:/clients/common/fdpp.git/build$ git diff
diff --git a/configure.meson b/configure.meson
index ab4d8666..ddd8b895 100755
--- a/configure.meson
+++ b/configure.meson
@@ -1,3 +1,8 @@
+#!/bin/sh
+
+
+set -xv
+
 if [ -f meson.build ]; then
     echo -e "Create a build dir first:"
     echo -e "\tmkdir build"
```

~~~
if [ -f meson.build ]; then
    echo -e "Create a build dir first:"
    echo -e "\tmkdir build"
    echo -e "\tcd build"
    echo -e "\t../$(basename $0)"
    echo "Bye!"
    exit 0
fi
+ [ -f meson.build ]

TOP=$(dirname "$0")
+ dirname ../configure.meson
+ TOP=..
meson setup --native-file $TOP/fdpp/toolchain.ini \
  --cross-file $TOP/fdpp/toolchain.ini \
  --cross-file $TOP/fdpp/kernel/toolchain.ini \
  --native-file $TOP/fdpp/kernel/toolchain.ini . $TOP --reconfigure
+ meson setup --native-file ../fdpp/toolchain.ini --cross-file ../fdpp/toolchain.ini --cross-file ../fdpp/kernel/toolchain.ini --native-file ../fdpp/kernel/toolchain.ini . .. --reconfigure
Directory does not contain a valid build tree:
/clients/common/fdpp.git/build
[ "$?" == "0" ] || exit 1
+ [ 1 == 0 ]
../configure.meson: 20: [: 1: unexpected operator
+ exit 1
~~~

**Comment by andrewbird** [comment]
so I think line 20 should be `[ "$?" = "0" ] || exit 1`  i.e. single `=`, but that's after  the error

**Comment by stsp** [comment]
Ok, not sure what does this mean.
From that build dir you can do:
```
$ find .. -name meson.build
../fdpp/thunk_gen/meson.build
../fdpp/kernel/meson.build
../fdpp/loader/meson.build
../fdpp/meson.build
../meson.build
```

Do you see the same?

**Comment by andrewbird** [comment]
this page https://mesonbuild.com/Configuring-a-build-directory.html suggests there's a step called `meson configure`

**Comment by stsp** [comment]
Please try to remove ` --reconfigure`
from script. Maybe in your version it
requires configured build tree.

**Comment by andrewbird** [comment]
yes that got further
~~~
ajb@calypso:/clients/common/fdpp.git/build$ ../configure.meson

if [ -f meson.build ]; then
    echo -e "Create a build dir first:"
    echo -e "\tmkdir build"
    echo -e "\tcd build"
    echo -e "\t../$(basename $0)"
    echo "Bye!"
    exit 0
fi
+ [ -f meson.build ]

TOP=$(dirname "$0")
+ dirname ../configure.meson
+ TOP=..
meson setup --native-file $TOP/fdpp/toolchain.ini \
  --cross-file $TOP/fdpp/toolchain.ini \
  --cross-file $TOP/fdpp/kernel/toolchain.ini \
  --native-file $TOP/fdpp/kernel/toolchain.ini . $TOP
+ meson setup --native-file ../fdpp/toolchain.ini --cross-file ../fdpp/toolchain.ini --cross-file ../fdpp/kernel/toolchain.ini --native-file ../fdpp/kernel/toolchain.ini . ..
The Meson build system
Version: 0.61.2
Source dir: /clients/common/fdpp.git
Build dir: /clients/common/fdpp.git/build
Build type: cross build
Project name: fdpp
Project version: 0.1
C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C linker for the host machine: gcc ld.bfd 2.38
C++ compiler for the host machine: clang++ (clang 14.0.0-1ubuntu1 "Ubuntu clang version 14.0.0-1ubuntu1.1")
C++ linker for the host machine: clang++ ld.bfd 2.38

../meson.build:1:0: ERROR: Tried to use unknown language "nasm".

A full log can be found at /clients/common/fdpp.git/build/meson-logs/meson-log.txt
[ "$?" = "0" ] || exit 1
+ [ 1 = 0 ]
+ exit 1
~~~

**Comment by stsp** [comment]
Added patch, can you please re-check?

**Comment by andrewbird** [comment]
Well that's the Jammy build stuffed then
~~~
ajb@calypso:/clients/common/fdpp.git/build$ ../configure.meson
The Meson build system
Version: 0.61.2
Source dir: /clients/common/fdpp.git
Build dir: /clients/common/fdpp.git/build
Build type: cross build

../meson.build:1:0: ERROR: Meson version is 0.61.2 but project requires >= 0.64

A full log can be found at /clients/common/fdpp.git/build/meson-logs/meson-log.txt
~~~

**Comment by stsp** [comment]
I am not quite sure things will improve
much... unless we want to drop everything
but 2024 LTS? :(
I guess we'll be able to switch CI to meson
at April, but seems like makefiles will have
to be supported in parallel for quite a years
to come. Which puts a big question mark
on this entire work.

**Comment by andrewbird** [comment]
is there an updated meson in a PPA we could use?

**Comment by stsp** [comment]
That's an interesting question.
In fact, as its written on python, we
probably can just use the Noble's
packages on any ubuntu.
Let me tell you secretly that I have
actually done exactly that. :) Just
forgot in 2 weeks of porting.

So... can you try to just to update to
the Noble build? IIRC there was no
such PPA, or why would I otherwise
use the Noble build myself.

**Comment by andrewbird** [comment]
~~~
ajb@calypso:/clients/common/fdpp.git/build$ sudo apt remove meson
~~~

~~~
ajb@calypso:/clients/common/fdpp.git/build$ sudo apt install ninja-build
[sudo] password for ajb: 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed
  ninja-build
0 to upgrade, 1 to newly install, 0 to remove and 0 not to upgrade.
Need to get 111 kB of archives.
After this operation, 358 kB of additional disk space will be used.
Get:1 http://gb.archive.ubuntu.com/ubuntu jammy/universe amd64 ninja-build amd64 1.10.1-1 [111 kB]
Fetched 111 kB in 0s (301 kB/s)       
Selecting previously unselected package ninja-build.
(Reading database ... 328447 files and directories currently installed.)
Preparing to unpack .../ninja-build_1.10.1-1_amd64.deb ...
Unpacking ninja-build (1.10.1-1) ...
Setting up ninja-build (1.10.1-1) ...
Processing triggers for man-db (2.10.2-1) ...
Processing triggers for doc-base (0.11.1) ...
Processing 1 added doc-base file...
~~~

~~~
ajb@calypso:/clients/common/fdpp.git/build$ pip3 install meson
Defaulting to user installation because normal site-packages is not writeable
Collecting meson
  Downloading meson-1.3.0-py3-none-any.whl (976 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 976.4/976.4 KB 5.0 MB/s eta 0:00:00
Installing collected packages: meson
  WARNING: The script meson is installed in '/home/ajb/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed meson-1.3.0
ajb@calypso:/clients/common/fdpp.git/build$ export PATH=/home/ajb/.local/bin:${PATH}
~~~
~~~
ajb@calypso:/clients/common/fdpp.git/build$ ../configure.meson
The Meson build system
Version: 1.3.0
Source dir: /clients/common/fdpp.git
Build dir: /clients/common/fdpp.git/build
Build type: cross build
Project name: fdpp
Project version: 0.1
Nasm compiler for the host machine: nasm-segelf (nasm 2.17)
Nasm linker for the host machine: gcc ld.bfd 2.38
C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C linker for the host machine: gcc ld.bfd 2.38
C++ compiler for the host machine: clang++ (clang 14.0.0-1ubuntu1 "Ubuntu clang version 14.0.0-1ubuntu1.1")
C++ linker for the host machine: clang++ ld.bfd 2.38
Nasm compiler for the build machine: nasm-segelf (nasm 2.17)
Nasm linker for the build machine: gcc ld.bfd 2.38
C compiler for the build machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C linker for the build machine: gcc ld.bfd 2.38
C++ compiler for the build machine: clang++ (clang 14.0.0-1ubuntu1 "Ubuntu clang version 14.0.0-1ubuntu1.1")
C++ linker for the build machine: clang++ ld.bfd 2.38
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86
Host machine cpu: i386
Target machine cpu family: x86
Target machine cpu: i386

Executing subproject libfdpp 

libfdpp| Project name: libfdpp
libfdpp| Project version: 0.1
libfdpp| C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
libfdpp| C linker for the host machine: gcc ld.bfd 2.38
libfdpp| C++ compiler for the host machine: clang++ (clang 14.0.0-1ubuntu1 "Ubuntu clang version 14.0.0-1ubuntu1.1")
libfdpp| C++ linker for the host machine: clang++ ld.bfd 2.38
libfdpp| C compiler for the build machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
libfdpp| C linker for the build machine: gcc ld.bfd 2.38
libfdpp| C++ compiler for the build machine: clang++ (clang 14.0.0-1ubuntu1 "Ubuntu clang version 14.0.0-1ubuntu1.1")
libfdpp| C++ linker for the build machine: clang++ ld.bfd 2.38

Executing subproject libfdpp:thunk_gen

thunk_gen| Project name: thunk_gen
thunk_gen| Project version: 0.1
thunk_gen| C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
thunk_gen| C linker for the host machine: gcc ld.bfd 2.38
thunk_gen| C compiler for the build machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
thunk_gen| C linker for the build machine: gcc ld.bfd 2.38
thunk_gen| Program flex found: YES (/usr/bin/flex)
thunk_gen| Program bison found: YES (/usr/bin/bison)
thunk_gen| Build targets in project: 1
thunk_gen| Subproject thunk_gen finished.

libfdpp| Fetching value of define "FDPP_API_VER" : 35
libfdpp| Fetching value of define "BPRM_VER" : 10
libfdpp| Program /clients/common/fdpp.git/subprojects/libfdpp/parsers/mkfar.sh found: YES (/clients/common/fdpp.git/subprojects/libfdpp/parsers/mkfar.sh)
libfdpp| Program /clients/common/fdpp.git/subprojects/libfdpp/parsers/parse_decls.sh found: YES (/clients/common/fdpp.git/subprojects/libfdpp/parsers/parse_decls.sh)
libfdpp| Program autom4te found: YES (/usr/bin/autom4te)
libfdpp| Program echo found: YES (/usr/bin/echo)
libfdpp| Build targets in project: 9
libfdpp| Subproject libfdpp finished.


Executing subproject loader 

loader| Project name: loader
loader| Project version: 0.1
loader| C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
loader| C linker for the host machine: gcc ld.bfd 2.38
loader| C compiler for the build machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
loader| C linker for the build machine: gcc ld.bfd 2.38
loader| Found pkg-config: YES (/usr/bin/pkg-config) 1.8.0
loader| Run-time dependency libelf found: YES 0.186
loader| Fetching value of define "FDPP_API_VER" : 35
loader| Fetching value of define "BPRM_VER" : 10
loader| Build targets in project: 10
loader| Subproject loader finished.


Executing subproject kernel 

kernel| Project name: kernel
kernel| Project version: 1.7
kernel| Nasm compiler for the host machine: nasm-segelf (nasm 2.17)
kernel| Nasm linker for the host machine: gcc ld.bfd 2.38
kernel| C compiler for the host machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
kernel| C linker for the host machine: gcc ld.bfd 2.38
kernel| Nasm compiler for the build machine: nasm-segelf (nasm 2.17)
kernel| Nasm linker for the build machine: gcc ld.bfd 2.38
kernel| C compiler for the build machine: gcc (gcc 11.4.0 "gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
kernel| C linker for the build machine: gcc ld.bfd 2.38
kernel| Fetching value of define "FDPP_API_VER" : 35
kernel| Fetching value of define "BPRM_VER" : 10
kernel| Program i686-linux-gnu-ld found: YES (/usr/bin/i686-linux-gnu-ld)
kernel| Build targets in project: 15
kernel| Subproject kernel finished.

Build targets in project: 15

fdpp 0.1

  Subprojects
    kernel      : YES
    libfdpp     : YES
    loader      : YES
    thunk_gen   : YES (from libfdpp)

  User defined options
    Cross files : ../fdpp/toolchain.ini
                  ../fdpp/kernel/toolchain.ini
    Native files: ../fdpp/toolchain.ini
                  ../fdpp/kernel/toolchain.ini

Found ninja-1.10.1 at /usr/bin/ninja

Done configure
Now run "meson build --verbose" to build
After that you can do "sudo meson install" to install
~~~



~~~
ajb@calypso:/clients/common/fdpp.git/build$ meson build --verbose
usage: meson [-h] {setup,configure,dist,install,introspect,init,test,wrap,subprojects,rewrite,compile,devenv,env2mfile,help} ...
meson: error: unrecognized arguments: --verbose
~~~

~~~
ajb@calypso:/clients/common/fdpp.git/build$ meson build

ERROR: Neither source directory 'build' nor build directory None contain a build file meson.build.
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
~~~

**Comment by stsp** [comment]
Wow!
Let me do just that...

**Comment by stsp** [comment]
I did.
Except that pip3 complained and asked
me to please use `pipx` instead. :)
As to your problem, try running `ninja`
as a work-around.
This meson, while matches my version
numerically, behaves very differently...

**Comment by andrewbird** [comment]
It should have been `meson compile`.
~~~
ajb@calypso:/clients/common/fdpp.git/build$ meson compile
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja
[17/122] Compiling C object subprojects/libfdpp/subprojects/thunk_gen/thunk_gen.p/meson-generated_thunk_gen.l.yy.c.o
subprojects/libfdpp/subprojects/thunk_gen/thunk_gen.p/thunk_gen.l.yy.c:1518:16: warning: ‘input’ defined but not used [-Wunused-function]
 1518 |     static int input  (void)
      |                ^~~~~
subprojects/libfdpp/subprojects/thunk_gen/thunk_gen.p/thunk_gen.l.yy.c:1475:17: warning: ‘yyunput’ defined but not used [-Wunused-function]
 1475 |     static void yyunput (int c, char * yy_bp )
      |                 ^~~~~~~
[122/122] Linking target subprojects/libfdpp/libfdpp.so.35.10
~~~

**Comment by stsp** [comment]
Yes, that's true.
Silly me. :)
Fixed.

**Comment by andrewbird** [comment]
I'll see if I can get it to work on CI now.

**Comment by stsp** [comment]
Please use `pipx` on CI to stay on the same
boat with me.

**Comment by stsp** [comment]
Fedora builds are no longer using pipx
for installing meson, as fedora-39 got
meson-1.3.2 via updates.
Whereas ubuntu builds are still on a
makefiles, as getting new meson even
in ubuntu-22.04 seems unlikely...

**Comment by volkertb** [comment]
@stsp Before opening a new ticket about this:

I can't build the most recent `master` of `fdpp`. Both `./configure` and `./configure.meson` complain about not finding `thunk_gen`, even though I had pulled that repo and had built it and installed it as well, using the commands `./configure ; meson compile --verbose -C build ; sudo meson install -C build`.

And I'm just building a new version of `fdpp` to satisfy the `dosemu2` build.

I'm sure you've had good reasons to make these changes, but it has gotten noticeably harder to build `dosemu2` from source than it used to be.

**Comment by stsp** [comment]
You simply have problems with
pkg-config. Namely, with PKG_CONFIG_PATH.
See `pkg-config --cflags thunk_gen`

**Comment by stsp** [comment]
> but it has gotten noticeably harder to build dosemu2 from source than it used to be.

But you can use ubuntu or fedora packages,
and also flatpak.
